components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -8979,4 +8982,110 @@ private void deleteScopes(Set<String> scopes, int tenantId) throws APIManagement
             }
         }
     }
+
+    /**
+     * Adds a new APIRevision to an existing API
+     *
+     * @param apiRevision APIRevision
+     * @throws APIManagementException if failed to add APIRevision
+     */
+    @Override
+    public String addAPIRevision(APIRevision apiRevision) throws APIManagementException {
+        int revisionId = apiMgtDAO.getMostRecentRevisionId(apiRevision.getApiUUID()) + 1;
+        apiRevision.setId(revisionId);
+        String revisionUUID = createAPIRevisionRegistryArtifacts(apiRevision.getApiUUID(), revisionId);
+        if (StringUtils.isEmpty(revisionUUID)) {
+            String errorMessage = "Failed to retrieve revision uuid";
+            log.error(errorMessage);
+            throw new APIManagementException(errorMessage);
+        }
+        apiRevision.setRevisionUUID(revisionUUID);
+        apiMgtDAO.addAPIRevision(apiRevision);
+        return revisionUUID;
+    }
+
+    /**
+     * Get a Revision related to provided and revision UUID
+     *
+     * @param revisionUUID API Revision UUID
+     * @return API Revision
+     * @throws APIManagementException if failed to get the related API revision
+     */
+    @Override
+    public APIRevision getAPIRevision(String revisionUUID) throws APIManagementException {
+        return apiMgtDAO.getRevisionByRevisionUUID(revisionUUID);
+    }
+
+    /**
+     * Copy Api registry artifacts to the revision registry location
+     *
+     * @param apiUUID API UUID
+     * @param revisionId Revision ID
+     * @throws APIManagementException if failed to copy API registry artifacts
+     */
+    protected String createAPIRevisionRegistryArtifacts(String apiUUID, int revisionId) throws APIManagementException {
+        String revisionUUID = null;
+        APIIdentifier apiId = APIUtil.getAPIIdentifierFromUUID(apiUUID);
```

Comment:
Shall we separate out the Dao layer accessing code from the registry accessing code. this way it would be easy to extract the registry code later

Reply:
separated the Dao and registry code, resolved